carcara/src/checker/rules/strings.rs

```diff
@@ -6,25 +6,31 @@ use super::{
 };
 use crate::ast::*;
 
-fn flatten(mut term: Rc<Term>, pool: &mut dyn TermPool) -> Vec<Rc<Term>> {
+fn flatten(term: Rc<Term>, pool: &mut dyn TermPool) -> Vec<Rc<Term>> {
     let mut flattened = Vec::new();
     if let Term::Const(Constant::String(s)) = term.as_ref() {
         flattened.extend(
             s.chars()
                 .map(|c| pool.add(Term::Const(Constant::String(c.to_string())))),
         );
-    } else {
-        while let Some((head, tail)) = match_term!((strconcat head tail) = term) {
-            if let Term::Const(Constant::String(s)) = head.as_ref() {
-                flattened.extend(
-                    s.chars()
-                        .map(|c| pool.add(Term::Const(Constant::String(c.to_string())))),
-                );
-            } else {
-                flattened.push(head.clone());
-            }
-            term = tail.clone();
+    } else if let Some(args) = match_term!((strconcat ...) = term) {
+        for arg in args {
+            match arg.as_ref() {
+                Term::Const(Constant::String(s)) => {
+                    flattened.extend(
+                        s.chars()
+                            .map(|c| pool.add(Term::Const(Constant::String(c.to_string())))),
+                    );
+                }
+                Term::Op(Operator::StrConcat, _) => {
+                    flattened.extend(flatten(arg.clone(), pool));
+                }
+                _ => {
+                    flattened.push(arg.clone());
+                }
```

Comment:
Notice that what you're doing in this `match` expression is basically what the body of `flatten` already does. So to simplify you can just call it directly:
```suggestion
            flattened.extend(flatten(arg.clone(), pool));
```